hazelcast/src/main/java/com/hazelcast/query/impl/IndexUtils.java

```diff
@@ -0,0 +1,349 @@
+     * Validate provided index config and normalize it's name and attribute names.
+     *
+     * @param mapName Name of the map
+     * @param config Index config.
+     * @return Normalized index config.
+     * @throws IllegalArgumentException If index configuration is invalid.
+     */
+    @SuppressWarnings("checkstyle:npathcomplexity")
+    public static IndexConfig validateAndNormalize(String mapName, IndexConfig config) {
+        assert config != null;
+
+        // Validate attributes.
+        List<String> originalAttributeNames = config.getAttributes();
+
+        if (originalAttributeNames.isEmpty()) {
+            throw new IllegalArgumentException("Index must have at least one attribute: " + config);
+        }
+
+        if (originalAttributeNames.size() > MAX_ATTRIBUTES) {
+            throw new IllegalArgumentException("Index cannot have more than " + MAX_ATTRIBUTES
+                + " attributes: " + config);
+        }
+
+        List<String> normalizedAttributeNames = new ArrayList<>(originalAttributeNames.size());
+
+        for (String originalAttributeName : originalAttributeNames) {
+            validateAttribute(config, originalAttributeName);
+
+            originalAttributeName = originalAttributeName.trim();
+
+            String normalizedAttributeName = canonicalizeAttribute(originalAttributeName);
+
+            assert !normalizedAttributeName.isEmpty();
+
+            int existingIdx = normalizedAttributeNames.indexOf(normalizedAttributeName);
+
+            if (existingIdx != -1) {
+                String duplicateOriginalAttributeName = originalAttributeNames.get(existingIdx);
+
+                if (duplicateOriginalAttributeName.equals(originalAttributeName)) {
+                    throw new IllegalArgumentException("Duplicate attribute name [attributeName="
+                        + originalAttributeName + ", indexConfig=" + config + ']');
+                } else {
+                    throw new IllegalArgumentException("Duplicate attribute names ["
+                        + "attributeName1=" + duplicateOriginalAttributeName + ", attributeName2="
+                        + originalAttributeName + ", indexConfig=" + config + ']');
+                }
+            }
+
+            normalizedAttributeNames.add(normalizedAttributeName);
+        }
+
+        // Construct final index.
+        String name = config.getName();
+
+        if (name != null && name.trim().isEmpty()) {
+            name = null;
+        }
+
+        return buildNormalizedConfig(mapName, config.getType(), name, normalizedAttributeNames);
+    }
+
+    private static IndexConfig buildNormalizedConfig(String mapName, IndexType indexType, String indexName,
+        List<String> normalizedAttributeNames) {
+        IndexConfig newConfig = new IndexConfig().setType(indexType);
+
+        StringBuilder nameBuilder = indexName == null
+            ? new StringBuilder(mapName + (indexType == IndexType.SORTED ? "_sorted" : "_hash")) : null;
```

Comment:
`indexType.name()`?

Reply:
I avoided that intentionally for the same reason why we have IDs in our configuration enums - even though it is unlikely that we will change index type names in the future. Currently, the name of the enum doesn't affect the name of the index, which is good IMO. WDYT?

Reply:
Agree, but `.name()` pattern is used in other places already, for instance in `getIndexTypeFromXmlName`. I'm fine with any solution, `indexType == IndexType.SORTED ? "_sorted" : "_hash"` part just looks fragile to me: if we would add a new index type, it will produce a hard to find/notice bug.

Reply:
Please see https://github.com/hazelcast/hazelcast/pull/15537/commits/2687ee15cf6ea98a01e91ba1c7553c38ae6cad89
Should be better now, WDYT?

Reply:
LGTM, thanks